core/trino-main/src/main/java/io/trino/server/security/HeaderBasedAuthenticatorConfig.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.server.security;
+
+import com.google.common.collect.ImmutableList;
+import io.airlift.configuration.Config;
+import io.airlift.configuration.ConfigDescription;
+import io.airlift.configuration.validation.FileExists;
+
+import javax.validation.constraints.NotEmpty;
+import javax.validation.constraints.NotNull;
+
+import java.io.File;
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class HeaderBasedAuthenticatorConfig
+{
+    private List<String> headers = ImmutableList.of("x-forwarded-client-cert");
```

Comment:
Configuring the list of headers globally for all authenticators doesn't seem right, since each authenticator will likely need separate headers. Also, exactly which headers are needed seems like a detail of the authenticator implementation, and not something the user configuring it would need or care to know.

It seems like we should avoid passing all headers to every authenticator, though I'm having trouble articulating why. That plus the case insensitivity issue makes using `Map` as the API problematic. I'm thinking we could do this for the API:
```java
public interface HeaderBasedAuthenticator
{
    interface Headers
    {
        // argument is treated as case insensitive
        List<String> getHeaders(String name);
    }

    Principal createAuthenticatedPrincipal(Headers headers);
}
```

This would open the door to adding a `getHeaderNames()` method later if needed. The implementation would be efficient since the lookup is lazy -- we don't need to copy the map before every call or try to delegate/wrap the existing map from JAX-RS.